streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java

```diff
@@ -45,24 +45,23 @@
     protected ProcessorNode<?, ?> currentNode;
     private long currentSystemTimeMs;
 
-    final StateManager stateManager;
```

Comment:
This is basically to avoid the cast from `StateManager` to `ProcessorStateManager` in `ProcessorContextImpl`. `ProcessorStateManager` has method `registeredChangelogPartitionFor()` that does only exist in the `ProessorStateManager`, because it is only used in `ProcessorContextImpl`. Hence, I moved the state manager field to the respective children and introduced the abstract method `stateManager()` in `AbstractProcessorContext`. Method `stateManager()` returns a `ProcessorStateManager` in `ProcessorContextImpl` but a `StateManager` in all other places.

Reply:
I see. Would it not be sufficient to just keep a ("duplicate") reference of `ProcessorStateManager` within `ProcessorContextImpl`? 

Just to clarify: I am ok with the proposed changes. Just wondering if it's really the best structure.

Reply:
If I understand you correctly, you propose to have a processor state manager reference in the `AbstractProcessorContext` and in `ProcessorContextImpl` instead of in the `GlobalProcessorContextImpl` and in `ProcessorContextImpl`. Moreover, you want to have a method `stateManager()` in `AbstractProcessorContext` that is overridden only in `ProcessorContextImpl`.

FWIW, I think it is cleaner to have the references in each child and an abstract method `stateManager()` in `AbstractProcessorContext` that is overridden in both children. My reasoning  is that both children have a state manager that is used in `AbstractProcessorContext` (i.e., both should provide a method `stateManager()`) but each child uses a different type of state manager internally.